configuration/builders/infra/runtime.py

```diff
@@ -0,0 +1,248 @@
+            result.append(
+                steps.ShellCommand(
+                    name=f"Prepare in container ({self.config.image_tag}) workdirs",
+                    command=util.Interpolate(
+                        (
+                            "docker run --rm "
+                            f"--mount type=volume,src={self.buildername},dst=/home/buildbot "
+                            f"-w {self.config.basedir} "
+                            f'{self.container_image} mkdir -p {" ".join(workdirs)}'
+                        )
+                    ),
+                    haltOnFailure=True,
+                )
+            )
+        for step in self.steps:
+            step.add_cmd_prefix(
+                [
+                    "docker",
+                    "run",
+                    "--init",  # To proper handle signals
+                    "--name",
+                    util.Interpolate(f"{self.buildername}"),
+                    "-u",
+                    f"{step.command.user}",
+                ]
+            )
+
+            if not hasattr(step, "checkpoint") or not step.checkpoint:
+                step.add_cmd_prefix(["--rm"])
+
+            for src, dst, type in self.config.volume_mounts:
+                step.add_cmd_prefix(
+                    [
+                        "--mount",
+                        util.Interpolate(f"type={type},src={src},dst={dst}"),
+                    ]
+                )
+            # Add env vars (global << sequence >> + local << step >>)
+            if step.env_vars:
+                env_vars = {env[0]: env[1] for env in self.config.env_vars}
+                env_vars.update({env[0]: env[1] for env in step.env_vars})
+                env_vars = list(env_vars.items())
+            else:
+                env_vars = self.config.env_vars
+
+            for env in env_vars:
+                step.add_cmd_prefix(["-e", util.Interpolate(f"{env[0]}={env[1]}")])
+
+            step.add_cmd_prefix([f"--shm-size={self.config.shm_size}"])
+
+            # Ignore basedir when an absolute path is given
+            if os.path.isabs(step.command.workdir):
+                step.add_cmd_prefix(["-w", f"{step.command.workdir}"])
+            else:
+                # (TODO: Razvan) No guarantee that this is a valid relative path
+                step.add_cmd_prefix(
+                    ["-w", f"{self.config.basedir}/{step.command.workdir}"]
+                )
+
+            step.add_cmd_prefix([self.container_image])
+
+            # User defined step to run in the container
+            result.append(step.generate())
+
+            # Create a checkpoint
+            if hasattr(step, "checkpoint") and step.checkpoint:
+                result.append(
+                    steps.ShellCommand(
+                        name=f"Checkpoint {step.name}",
+                        command=[
+                            "bash",
+                            "-c",
+                            util.Interpolate(
```

Comment:
could use `docker container commit --message {step.name}` to aid debugging/further activities.